README.md
# MyWallet V3 Frontend [![Build Status](https://travis-ci.org/blockchain/My-Wallet-V3-Frontend.png?branch=master)](https://travis-ci.org/blockchain/My-Wallet-V3-Frontend) [![Coverage Status](https://coveralls.io/repos/blockchain/My-Wallet-V3-Frontend/badge.svg?branch=master&service=github)](https://coveralls.io/github/blockchain/My-Wallet-V3-Frontend?branch=master)

An AngularJS bitcoin web wallet powered by [My-Wallet-V3](https://github.com/blockchain/My-Wallet-V3).

This is the new and improved wallet. You can see it at [alpha.blockchain.info](https://alpha.blockchain.info/). For the original wallet at [blockchain.info](https://blockchain.info/) please see [this repository](https://github.com/blockchain/My-Wallet) or [contact support](http://blockchain.zendesk.com/).

## Setup

Make sure you have [Node.js](http://nodejs.org/) installed.

Some NodeJS components need to be installed system wide:
```sh
npm install -g grunt-cli coffee-script
```

You also need Sass (use `sudo` if you're not using a [Ruby version manager](https://rvm.io)):
```sh
gem install sass
```

Install dependencies:
```sh
npm install
```

Create a file called `.env` in the root of the project. Put the following in it:

```
ROOT_URL=https://blockchain.info
```

## Build

Grunt watches and compiles the Jade view templates and CSS. Keep it running:
```sh
grunt
```

## Test

To run test and monitor for changes:
```sh
npm test
```

A coverage report is generated after you run the test for the first time. Just open `coverage/index.html` in your browser.

## Run

Run local http server:
```sh
npm start
```

Visit [local.blockchain.com:8080](http://local.blockchain.com:8080/).  Do not use `localhost:8080`. You will need to modify your "hosts" file (`/etc/hosts` on OSX and most UNIX systems) because this is no longer registered at the DNS level for application security reasons. Add this line to `/etc/hosts`:

    127.0.0.1   local.blockchain.com

## Developing My-Wallet-V3

If you are making changes to [My-Wallet-V3](https://github.com/blockchain/My-Wallet-V3) that you want to try out in the frontend, create a symlink:
```sh
rm My-Wallet-V3-Frontend/bower_components/blockchain-wallet/dist/my-wallet.js
ln -s ../../../../My-Wallet-V3/dist/my-wallet.js My-Wallet-V3-Frontend/bower_components/blockchain-wallet/dist/my-wallet.js
```

## Usage

You can open any wallet registered with your email address. It will ask you to upgrade to HD if needed. You can also sign up for a new wallet.

After login, you'll see a list of accounts. There will be delay before transactions and the correct balances show up. If something goes wrong during the login process, error messages appear in the console.

To automatically login and go back to where you last were in the app after every page refresh, create a file `.env` and add `AUTO_RELOAD=1` to it.

To reclaim funds from an email take code at the of link and add it to `#/claim`, e.g.:

    https://blockchain.info/wallet/claim#7Educ5YNnVPQCQ556w7W8tQpj1dchhxPK56vVNab68cK
    http://local.blockchain.com:8080/#/claim/7Educ5YNnVPQCQ556w7W8tQpj1dchhxPK56vVNab68cK

If you enable "handle bitcoin links" in your wallet settings, you can open bitcoin URI's like this one:

    bitcoin:?address=1FeerpCgswvGRLVKme759C96DUBtf7SvA2?amount=0.01

## Contribute

Did you know you can [sign your commits](https://git-scm.com/book/tr/v2/Git-Tools-Signing-Your-Work) using a PGP key?

## Testnet

Not supported by the server yet.

## Deploy

Create a static HTML/JS/CSS distribution package in `dist`.

    grunt dist

If you get 403 error from Github (because you exceeded their rate limit), create a [personal access token](https://github.com/settings/tokens). Only select `public_repo` from the list.

    GITHUB_USER=... GITHUB_TOKEN=... grunt dist

If you don't care about securely downloading dependencies and want to avoid using your Github credentials, use `grunt dist_unsafe` instead.

You can test the resulting files by setting `DIST=1` in `.env` and restarting the server.

`index.html` should be cached using `If-Modified-Since` or `etag`. All other files contain a hash of their content and should be cached forever.

## Security

Security issues can be reported to us in the following venues:

 * Email: security@blockchain.info
 * Bug Bounty: https://www.crowdcurity.com/blockchain-info


Changelog.md
__Blockchain HD Frontend__

_Recent changes_

#   (2016-01-29)



---

## Bug Fixes

- **2FA:** don't allow phone number change while SMS 2FA is enabled
  ([9bf36310](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/9bf36310924ac381610f3ace924dc778ab40887a))
- **DidYouKnow:** use states instead of paths, cleaned up and added tests
  ([7d498725](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/7d498725ebe7bbe58ff6ced867f579efadaa21d7))
- **SecurityCenter:** email and password hint interaction with bcasync
  ([f8ac727c](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/f8ac727c36d0b0e9f9613479c6a6deb5acc08fb7))
- **Upgrade:** don't allow ESC key to close modal
  ([6b4f9caf](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/6b4f9caf8debafb12e58c85f5ca11ca19495053b))
- **activityFeed:** update recent transactions when display currency is toggled
  ([cd9ef576](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/cd9ef5765a34f2f87649dec43d81dce5af7f4f0c))
- **alerts:** use directive to display alerts, fix closing behavior
  ([7cf31d37](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/7cf31d37e9bf9e1d0237b98ae63fbaeb523d0c11))
- **bcAsyncInput:** interaction fix for hiding and showing input fields
  ([640dbb1a](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/640dbb1a4eb53827e6d5ba85ae5e4b5b9d77cffe))
- **changePassword:**
  - do not allow changing to current password
  ([31a2a880](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/31a2a88086263fc49a9588d75797fb9d17d989b0))
  - do not allow user to set email as wallet password
  ([52e6ca0f](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/52e6ca0fb682d9719113ab23002a6cc4f5ba22d3))
- **home:** show imported addresses in balance table even with 0 balance
  ([d6f0eb12](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/d6f0eb12dd09c9e0a621d9f7896fb535bbda2664))
- **note:** trigger update after setting a tx note
  ([dd1359e1](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/dd1359e14077e0d91f6afd7750654d4626d69345))
- **preferences:** add minutes unit to logout time display
  ([b08ebe1b](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/b08ebe1b9a4ae0868e9697bc8ecf3a475f8ace28))
- **security:**
  - escape html where ng-bind-html is used
  ([bca8e00c](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/bca8e00c6c0245460fd954f5f53850ce90678bb7))
  - show error when enable ip restrict fails
  ([9b05579f](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/9b05579fe46747af3a546689557ab05a0d81a397))
- **send:**
  - pass payment through error
  ([96e9955d](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/96e9955d5a632d8cc1e412b44981907c059e231d))
  - rebuild tx after send failure so that retrying is possible
  ([9bad693f](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/9bad693fbe101f01d01e71e1f8125884cd56c08e))
- **translations:** fix display of translations followed by :
  ([db89bb9d](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/db89bb9d74f297d001dcf27cd9cdd576953011fe))


## Features

- **Deploy:** hardcode version and show in console
  ([e9ac4c38](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/e9ac4c38c490b0882bc8c4ff4476adb930d67bc9))
- **Upgrade:** redirect to classic login if refused
  ([5b267792](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/5b26779298415883d802ecbdf83959c386461189))
- **dev:**
  - replace savePassword with autoReload
  ([3cd6513a](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/3cd6513add54ed690a21c2076b3424dcd07ef9ca))
  - automatically login and go to last url when flag is set
  ([794284ec](https://github.com/blockchain/My-Wallet-HD-Frontend/commit/794284ec8d6ccb5e27471633ea744c6460ed7b11))


## Test

- **SecurityCenter:** add tests
- **changePassword:** test for setting password to email / current password



---
<sub><sup>*Generated with [git-changelog](https://github.com/rafinskipg/git-changelog). If you have any problem or suggestion, create an issue.* :) **Thanks** </sub></sup>

server.js
'use strict';

var express         = require('express')
  , ejs             = require('ejs')
  , path            = require('path')

loadEnv('.env');

var port      = process.env.PORT || 8080
  , dist      = parseInt(process.env.DIST) === 1
  , origins   = (process.env.BLOCKCHAIN || '').split(' ')
  , whitelist = (process.env.IP_WHITELIST || '').split(' ')
  , rootURL   = process.env.ROOT_URL || 'https://blockchain.info/'
  , webSocketURL = process.env.WEBSOCKET_URL || false

// App configuration
var app = express();

app.use(function (req, res, next) {
  if (req.url === '/') {
    var cspHeader = ([
      "img-src 'self' " + rootURL + " data:",
      "style-src 'self' 'sha256-vv5i1tRAGZ/gOQeRpI3CEWtvnCpu5FCixlD2ZPu7h84=' 'sha256-47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU='",
      "child-src 'none'",
      "script-src 'self' 'sha256-mBeSvdVuQxRa2pGoL8lzKX14b2vKgssqQoW36iRlU9g=' 'sha256-47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU='",
      "connect-src 'self' " + rootURL + " " + (webSocketURL || "wss://*.blockchain.info") + " https://api.blockchain.info",
      "object-src 'none'",
      "media-src 'self' data: mediastream: blob:",
      "font-src 'self'", ''
    ]).join('; ');
    res.setHeader('content-security-policy', cspHeader);
    res.setHeader('X-Frame-Options', 'SAMEORIGIN');

    var ip = req.headers['x-forwarded-for'] || req.connection.remoteAddress;
    if (whitelist != '' && whitelist.indexOf(ip.split(', ')[0]) < 0) {
      console.log(ip);
      res.status(403).send('I\'m sorry Dave, I can\'t let you do that.');
    } else if (dist) {
      res.render('index.html');
    } else {
      res.render('app/index.jade');
    }
    return;

  }
  if (dist) {
    res.setHeader('Cache-Control', 'public, max-age=31557600');
  } else {
    res.setHeader('Cache-Control', 'public, max-age=0, no-cache');
  }
  next();
});

if (dist) {
  console.log('Production mode: single javascript file, cached');
  app.engine('html', ejs.renderFile);
  app.use(express.static('dist'));
  app.set('views', path.join(__dirname, 'dist'));
} else {
  console.log('Development mode: multiple javascript files, not cached');
  app.use(express.static(__dirname));
  app.set('view engine', 'jade');
  app.set('views', __dirname);
}

app.use(function (req, res) {
  res.status(404).send('<center><h1>404 Not Found</h1></center>');
});

app.listen(port, function () {
  console.log('Listening on %d', port);
});

// Custom middleware
function allowOrigins(origins) {
  return function (req, res, next) {
    origins.forEach(function (origin) {
      if (req.headers.origin != null && req.headers.origin.indexOf(origin) > -1) {
        res.setHeader('Access-Control-Allow-Origin', req.headers.origin);
      }
    });
    next();
  };
}

// Helper functions
function loadEnv(envFile) {
  try {
    require('node-env-file')(envFile);
  } catch (e) {
    console.log('You may optionally create a .env file to configure the server.');
  }
}
